inc/class-contributor.php

```diff
@@ -0,0 +1,235 @@
+<?php
+/**
+ * @author  Pressbooks <code@pressbooks.com>
+ * @license GPLv2 (or any later version)
+ */
+
+namespace Pressbooks;
+
+use function Pressbooks\Utility\str_starts_with;
+
+class Contributor {
+
+	/**
+	 * Valid contributor slugs
+	 *
+	 * @var array
+	 */
+	public $valid = [
+		'pb_authors',
+		'pb_editors',
+		'pb_translators',
+		'pb_proofreaders',
+		'pb_reviewers',
+		'pb_illustrators',
+		'pb_contributors',
+	];
+
+	public function __construct() {
+	}
+
+	/**
+	 * Retrieve all author/editor/etc lists for a given Post ID
+	 *
+	 * @param int $post_id
+	 *
+	 * @return array
+	 */
+	public function getAll( $post_id ) {
+		$contributors = [];
+		foreach ( $this->valid as $contributor_type ) {
+			$contributors[ $contributor_type ] = $this->get( $post_id, $contributor_type );
+		}
+		return $contributors;
+	}
+
+	/**
+	 * Retrieve author/editor/etc lists for a given Post ID and Contributor type
+	 *
+	 * @param int Post ID or object.
+	 * @param string $contributor_type
+	 *
+	 * @return string
+	 */
+	public function get( $post_id, $contributor_type ) {
+		if ( ! str_starts_with( $contributor_type, 'pb_' ) ) {
+			$contributor_type = 'pb_' . $contributor_type;
+		}
+		if ( ! in_array( $contributor_type, $this->valid, true ) ) {
+			return '';
+		}
+
+		// Look if contributors exist as taxonomies (new data model)
+		$contributors = [];
+		$meta = get_post_meta( $post_id, $contributor_type, false );
+		foreach ( $meta as $slug ) {
+			$term = get_term_by( 'slug', $slug, 'contributor' );
+			if ( $term ) {
+				$first_name = get_term_meta( $term->term_id, 'contributor_first_name', true );
+				$last_name = get_term_meta( $term->term_id, 'contributor_last_name', true );
+				if ( ! empty( $first_name ) && ! empty( $last_name ) ) {
+					$contributors[] = "{$first_name} {$last_name}";
+				} elseif ( ! empty( $term->name ) ) {
+					$contributors[] = $term->name;
+				}
+			}
+		}
+
+		if ( empty( $contributors ) ) {
+			// Look if contributors exist as metadata (old data model)
+			// If yes then convert to taxonomies (new data model)
+			$map = [
+				'pb_authors' => [ 'pb_author', 'pb_section_author' ],
+				'pb_editors' => [ 'pb_editor' ],
+				'pb_translators' => [ 'pb_translator' ],
+				'pb_contributors' => [ 'pb_contributing_authors' ],
+			];
+			if ( isset( $map[ $contributor_type ] ) ) {
+				foreach ( $map[ $contributor_type ] as $slug ) {
+					$meta = get_post_meta( $post_id, $slug, false );
+					foreach ( $meta as $name ) {
+						$result = $this->insert( $name );
+						if ( $result !== false ) {
+							$added = $this->link( $result['term_id'], $post_id, $contributor_type );
+							if ( $added !== false ) {
+								$contributors[] = $name;
+							}
+						}
+					}
+				}
+			}
+		}
+
+		// Comma separated, Oxford comma, localized and between the last two items
```

Comment:
This would probably be good to turn into a function in `\Pressbooks\Utility\` (?) so that we can make a formatted, comma separated list with Oxford commas out of any array of items.